app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/DefaultTxExecutorRepository.kt

```diff
@@ -1,53 +1,109 @@
         billingRepository.observePurchases()
-            .map { it.contains(EXECUTOR_SERVICE_SUBSCRIPTION_ID) }
+            .map {
+                it.find { it.productId == EXECUTOR_SERVICE_CREDITS_ID }?.token?.let {
+                    DataResult(it)
+                } ?:
+                ErrorResult<String>(NoSuchElementException())
+            }
 
-    override fun buyPlan(activity: Activity): Single<Boolean> =
-        billingRepository.launchSubscribeFlow(activity, EXECUTOR_SERVICE_SUBSCRIPTION_ID)
+    override fun redeemVoucher(voucher: String): Single<Long> =
+        authentication()
+            .flatMap { (account, signature) ->
+                executorApi.reedeemVoucher(account, signature, TxExecutionVoucherData(voucher)).map { it.balance }
+            }
+            .onErrorResumeNext { throwable: Throwable ->
+                if (throwable is HttpException && throwable.code() == 409) Single.just(-1)
+                else Single.error(throwable)
+            }
+            .flatMap { balance ->
+                billingRepository.consume(voucher).map { balance }
+            }
 
     override fun execute(transaction: Transaction): Observable<String> =
-        billingRepository.checkSubscription(EXECUTOR_SERVICE_SUBSCRIPTION_ID)
-            .flatMapObservable {
-                it.toNullable()?.let {
-                    executorApi.executeTx(
-                        it.token,
-                        TxExecutionData(
-                            transaction.address.asEthereumAddressString(),
-                            transaction.data
-                        )
+        authentication()
+            .flatMapObservable { (account, signature) ->
+                executorApi.executeTx(
+                    account,
+                    signature,
+                    TxExecutionData(
+                        transaction.address.asEthereumAddressString(),
+                        transaction.data
                     )
-                } ?: Observable.error(IllegalStateException("No subscription"))
+                )
             }
             .map { it.hash }
 
+    override fun estimate(transaction: Transaction): Observable<Pair<Long, Long>> =
+        authentication()
+            .flatMapObservable { (account, signature) ->
+                executorApi.estimateTx(
+                    account, signature,
+                    TxExecutionData(
+                        transaction.address.asEthereumAddressString(),
+                        transaction.data
+                    )
+                )
+            }
+            .map { it.balance to it.requiredCredits }
+
+    private fun authentication(): Single<Pair<String, String>> =
+        accountsRepository.loadActiveAccount()
+            .flatMap {
+                val addressString = it.address.asEthereumAddressString()
+                accountsRepository.sign(Sha3Utils.keccak(addressString.hexStringToByteArray())).map { addressString to it.toString() }
```

Comment:
Why do we use `keccak` here?

Reply:
Need to reformulate the question: why use any hashing function here? 😉

Reply:
Ahhh ok, that is just a common pattern. You always sign a hash and not the real payload. E.g. when you sign a transaction you sign the hash of the rlp encoded transaction, or when you sign a message for the Gnosis Safe you sign the transaction hash (`keccak256(byte(0x19), this, to, value, data, operation, _nonce)`). I am not too deep into cryptography to give you the 100% correct answer why that is a good pattern 😄

Reply:
https://crypto.stackexchange.com/questions/12768/why-hash-the-message-before-signing-it-with-rsa

Reply:
https://en.wikipedia.org/wiki/Digital_signature#Method